src/main/scala/au/com/agiledigital/toolform/command/generate/docker/GenerateDockerComposeV3Command.scala

```diff
@@ -61,29 +63,42 @@ object GenerateDockerComposeV3Command extends YamlWriter {
     *                               error object describing what went wrong.
     */
   def runGenerateDockerComposeV3(sourceFilePath: String, outFile: File, project: Project): Either[NonEmptyList[ToolFormError], String] = {
-    val writer = new BufferedWriter(new FileWriter(outFile, false))
-    try {
-      val writeFile = for {
-        _ <- write(s"# Generated by ${BuildInfo.name} (${BuildInfo.version})")
-        _ <- write(s"# Source file: $sourceFilePath")
-        _ <- write(s"# Date: ${DateUtil.formattedDateString}")
-        _ <- write("version: '3'")
-        _ <- write("services:")
-        _ <- indented {
-              for {
-                _ <- writeComponents(project.id, project.sortedComponents.values.toList)
-                _ <- writeResources(project.sortedResources.values.toList)
-              } yield ()
-            }
-      } yield ()
-
-      val context = WriterContext(writer)
-      // Final state needs to be read for anything to happen because of lazy evaluation
-      val _ = writeFile.run(context).value
+    // validated image names
+    val imageNameErrors = project.sortedResources.values.toList.traverseU { resource =>
```

Comment:
Almost there! I just have some suggestions to restructure it a little bit and I'm happy to approve.

That is a great use of `traverseU` to fold down all the validation objects into a single object. I'd probably just suggest that we move the validation logic into its own method. Just to make it cleaner if we start adding more validation in future PRs. So you would end up with a separate method:

	  private def validateResource(resource: Resource): Validated[NonEmptyList[ToolFormError], Resource] =
	    if (resource.image.isEmpty)
	      invalid(NonEmptyList.of(ToolFormError(s"Image name is required for: ${resource.id}")))
	    else
	      valid(resource)

and you would just pass it into the `traverseU` call

	project.sortedResources.values.toList.traverseU(validateResource).toEither

After you have your validated resources of type `Validated[NonEmptyList[ToolFormError], List[Resource]]` you currently do a match on it so that you can execute the writer if the validation succeeds or just pass the error along if there is an error.

Instead of having to write `case Left(error) => Left(error)` every time there is a nicer way of doing that in Scala. If you call `map` and `flatMap` on an `Either` you can specify some code that will run if the either is successful but will be ignored if it has failed. For example, the code below chains together three methods that return an `Either`. If all three succeed then the final result will be a tuple of all the results. If one fails then the calls nested below it will never fire and it will skip straight to printing the error.

	// Written off the top of my head, could be wrong ;-)
	def finalResult = someOperationThatMightFail().flatMap { result => 
		someOtherOperationThatMightFail().flatMap { result2 =>
			aThirdOperationThatMightFail().map { result3 =>
				Right((result, result2, result3))
			}
		}
	}

	finalResult match {
       case Left(error) => println("Oh no an error: ${error}")
       case Right(result, result2, result3) => println("${result}, ${result2}, ${result3}")
    }

This is a really ugly pyramid of doom, so Scala gives us for comprehensions as syntactic sugar to clean this up. Using a for comprehension you can rewrite the `map` and `flatMap` statements to:

	def finalResult = for {
		result <- someOperationThatMightFail()
		result2 <- someOtherOperationThatMightFail()
		result3 <- aThirdOperationThatMightFail()
	} yield (result, result2, result3)

You can use for comprehensions on lots of types in Scale. Cats even defines `map` and `flatMap` on lots of its types which is why we can use the for comprehensions do run all the writer commands.

Also the resources are being validated but the result is getting thrown away in the `case Right(_) => {` statement. I think might be better to get the validated result and pass it into the writer, so we know that the resources that we pass in are definitely valid (although the type contract doesn't guarantee it at this point but I think that is something we can tackle later).

I suggest breaking out the writer code into another method called something like `writeAll` that takes a `List[Resource]` along with the other parameters passed into `runGenerateDockerComposeV3`

	private def writeAll(validResources: List[Resource], sourceFilePath: String, outFile: File, project: Project): Either[NonEmptyList[ToolFormError], String] = {
		// Writer code
	}

then changing the call to `writeResources` to use the validResources that are passed in
	
	_ <- writeResources(validResources)

and finally, changing the body of `runGenerateDockerComposeV3` to a for comprehension that:

1. Runs the resource validation and stores the validated resources
2. Runs `writeAll` passing in the validated resources
3. Returns the final status

It should look something like this:

	for {
      validatedResources <- project.sortedResources.values.toList.traverseU(validateResource).toEither
      writerStatus       <- writeAll(validatedResources, sourceFilePath, outFile, project)
    } yield writerStatus

As a final thing, because we now know that the resources passed into the writers are valid we could change

	resource.image.getOrElse(nil)

to just:

	resource.image.get

It will blow up if there is no image specified but it should never happen because we make sure we only pass in validated resources.

I hope that makes sense.

See also: https://docs.scala-lang.org/tutorials/FAQ/yield.html

Reply:
Thanks a lot for this comment. It was very easy to follow your steps and I can definitely see the improvement of writing it in this way 😄